### ISSUE TITLE
Upgrade from 3.5.5 to 3.6.0+ run very slowly

### ISSUE THREAD

**Issue by pmancaux** [liquibase/liquibase#1171]
## Environment
**Liquibase Version**: 3.5.5

**Liquibase Integration & Version**:  maven

**Liquibase Extension(s) & Version**: 

**Database Vendor & Version**: Postgresql 9

**Operating System Type & Version**: Ubuntu

## Description
When i upgrade liquibase from 3.5.5 to 3.6.0 or 3.9.0 my database changelog run very very slowly

In 3.5.5: 
INFO 27/05/20 12:16: liquibase: changelog-master-postgresql: classpath:db/changelogs/db.changelog-2019-02-07_12-00-DAHU-2777-postgresql.xml::1549537637247-7::lla (generated): ChangeSet classpath:db/changelogs/db.changelog-2019-02-07_12-00-DAHU-2777-postgresql.xml::1549537637247-7::lla (generated) ran successfully in 43ms

In 3.6.0+
12:13:54.888 [INFO ] [localhost-startStop-1] liquibase.changelog.ChangeSet - ChangeSet classpath:db/changelogs/db.changelog-2019-02-07_12-00-DAHU-2777-postgresql.xml::1549537637247-7::lla (generated) ran successfully in 18488ms

As you can see for this example i pass from 43ms to 18488ms and it's the same for all my changelog, basicaly i pass from 20sec for initialize my empty db to 20minutes!

I already open à Jira about this but no response : https://liquibase.jira.com/browse/CORE-3478

## Steps To Reproduce
To reproduce i can give you all my changelog but it's private project, i can work to anonymize all script and give you all if it's necessary

Thanks

**Comment by molivasdat** [comment]
Hi @pmancaux Thanks for taking the time to write this up again.  What liquibase command are you running to see these timings?  Performance issues are a little more difficult to track down since there are many reasons for them.  What types of change sets are taking the longest?

**Comment by pmancaux** [comment]
@molivasdat hello, sorry for time to respond,

to complete my issue :

After debugging few hours. Slowness seems to be related to ColumnSnapshotGeneratorPostgres  which is applied even when we are checking index presence.

ColumnSnapshotGeneratorPostgres overrides the getPriority method in charge of returning if generator should be used.  There is no control to check if the object type we want to snapshot is assignable from Column type.
public class ColumnSnapshotGeneratorPostgres extends ColumnSnapshotGenerator {



```
    @Override

    public int getPriority(Class<? extends DatabaseObject> objectType, Database database) {

        if (database instanceof PostgresDatabase)

            return PRIORITY_DATABASE;

        else

            return PRIORITY_NONE; // Other DB? Let the generic handler do it.

    }
```

  

...

}

Is there any reason why there is no control on type ?

**Comment by molivasdat** [comment]
Hi @pmancaux Thanks for pointing to the code that is the problem in your case?  Do you have a simple testcase that shows the problem?  What command are you running when this is happening?

**Comment by pmancaux** [comment]
@molivasdat hello, i m in spring application, i will try to create a simple test case to reproduce the comportment and i send you this rapidly. My test case look like this:
-create a postgres db with schema
-create many table (100 tables)
-create liquibase script with many not condition
-try this in 3.5.5 to see that script is executed rapidly
-try same scenari in 3.6.0+ to see that script is executed very slowly

i try to prepapre this soon

**Comment by pmancaux** [comment]
hello,

i have create a zip file with all you need to reproduce, you juste need to have a postgresql db installed or a dockerized one in version 9 or 12. after just do the following command:

create a database and schema testschema like 
ceatedb  -U postgres -h localhost liquibasetest
create SCHEMA testschema ;

run liquibase commande first with : 

java    -jar  liquibase-core-3.5.5.jar --driver=org.postgresql.Driver --classpath=postgresql-42.2.16.jar --changeLogFile=db/changelogs/db.changelog-master-postgresql.xml --url="jdbc:postgresql://localhost:5432/liquibasetest" --username=turtle --password=turtle --defaultSchemaName=testschema  --logLevel=debug update

DROP SCHEMA testschema cascade;
create SCHEMA testschema ;

and run a second time with liquibase 4.1.0 :
java    -jar  liquibase-core-4.1.0.jar --driver=org.postgresql.Driver --classpath=postgresql-42.2.16.jar --changeLogFile=db/changelogs/db.changelog-master-postgresql.xml --url="jdbc:postgresql://localhost:5432/liquibasetest" --username=turtle --password=turtle --defaultSchemaName=testschema  --logLevel=debug update

you will see that with the first command all is run in about ~40seconds and the second command in about 5 minutes!

and when you have multiple schema is very horrible to wait... please find a solution because i m stuck in liquibase 3.5.5


[liquibase-issue-1171.zip](https://github.com/liquibase/liquibase/files/5351108/liquibase-issue-1171.zip)

**Comment by pmancaux** [comment]
@molivasdat hello, did you have test my project to see the problem? did you need anything else?

thanks

**Comment by molivasdat** [comment]
Hi @pmancaux Thanks for the excellent recreation scenario.  Very well documented.  I just tried to recreate the issue just as you described and I was successful in recreating the issue.  I used postgres 11 because that is what I had available at the time I tried to recreate and it still showed the same overall results.  I'll get with the team and see what we can do to move you up to a later version.

**Comment by pmancaux** [comment]
@molivasdat just for any news if you have :) thanks!

**Comment by shaerinck-truefit** [comment]
Hi @molivasdat.  My company is also running into this issue when attempting to upgrade from 3.4.0 to 4.4.0.  Are there any updates on timing for a potential fix for this?  We have 1 build that went from 15 minutes on 3.4.0 to over 9 hours on 4.4.0 before I killed it.  This will clearly be a blocker for us to be able to upgrade.

**Comment by nvoxland** [comment]
Running your example with the current 4.4.0 code, it appears the #1952 solves the performance problem I'm still seeing.

**Comment by pmancaux** [comment]
@nvoxland I try to run my example in zip file with 4.4.0 and 4.4.1 but problem still exist.
Try to run with 3.5.5 and you see that all liquibase xml file are treated in  about 30 sec and in 4.4.0+ it never finish after multiple minutes...like in 3.6.0+, for me the problem still persist

**Comment by shaerinck-truefit** [comment]
@nvoxland I also pulled down 4.4.2 which I believe contains your fix but am still seeing performance issues. Your change definitely helped bring the time down from 9+ hours on 4.4.1 to 40 minutes for my build but 40 minutes is still a 3x performance decrease over using 3.4.0.

**Comment by nvoxland** [comment]
There is a separate performance fix that should be going into 4.4.3 which should help as well. 

After 4.4.3 comes out, @shaerinck-truefit , can you see if that solves the problem? 

If not, let me know and I can help you get a profile dump so we can understand where your problem is at for sure.
